### DISCUSSION TITLE
Standardize how to flag OS support in PowerShell Gallery packages, PSResourceGet and in PowerShell itself

### DISCUSSION THREAD

**Discussion by o-l-a-v** [PowerShell/PowerShell#24723] [Enhancements]
Edit: PSGallery currently have the option to optionally tag packages with `linux`, `mac` etc. for **discoverability**. This discussion is about making flagging and validating OS support a runtime behavior for consistency and user friendliness.

---

AFAIK there is no standardized way to flag what OS-es a PowerShell Gallery package supports. 

Some things to think about:

* Should support be flagged by tags, as some do today? `unix`, `linux`, `mac`, `windows`.
* Should it be up to the author to flag this correctly?
* Should the package registry perform some tests then apply some info about what OS-es is supported?
* Should PSResourceGet block installing a `windows` only package on a `unix` / `linux` / `mac` system, at least show a warning?
  * Currently `Install-PSResource -Repository 'PSGallery' -Name 'Microsoft.WinGet.Client'` on Linux gives no warnings or errors.
* Should `Import-Module` throw if trying to import a Windows only module on MacOS or Linux?
  * Currently `Import-Module -Name 'Microsoft.WinGet.Client'` on Linux gives no warnings or errors.

As @JustinGrote wrote here: <https://github.com/PowerShell/PSResourceGet/issues/1570>

> It's a fair point that just because a module is written for PSCore doesn't mean it supports all platforms. For instance, a module might be a wrapper to a windows-only or linux-only binary.

Relevant previous discussions / issues / feature requests:

* 2017-05-10: <https://github.com/PowerShell/PowerShell/issues/3751>
* 2018-05-23: <https://github.com/PowerShell/PowerShell-RFC/pull/122>
  * Conclusion: <https://github.com/PowerShell/PowerShell-RFC/pull/122#issuecomment-460419644>
* 2019-05-31: <https://github.com/PowerShell/PowerShell/issues/9784>
* 2019-07-29: <https://github.com/PowerShell/PowerShell/discussions/15502>

---

I think flagging and validating OS support should be standardized, documented and applied to all parts of PowerShell that handles packages (modules, scripts).

**Comment by rhubarb-geek-nz** [comment]
This sounds like a good idea. Currently the only related field in a module manifest is for the [processor architecture](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_module_manifests?view=powershell-7.4#processorarchitecture).

I recommend that the naming should follow the existing [dotnet RID catalog](https://learn.microsoft.com/en-us/dotnet/core/rid-catalog) naming and conventions. So for macOS that would be "osx" and Windows would be "win".

**Comment by ThomasNieto** [comment]
The PSGallery uses the Windows, Linux, and MacOS tags to filter while searching.

![image](https://github.com/user-attachments/assets/e6d957c9-e9f5-4980-a317-3137e5d0c538)

**Reply by rhubarb-geek-nz** [reply]
> The PSGallery uses the Windows, Linux, and MacOS tags to filter while searching.

How do you encode the operating system in the module?

**Reply by ThomasNieto** [reply]
In the module manifest tags property.

```powershell
@{
  PrivateData = @{
    PSData = @{
      Tags = @('Linux', 'Windows')
    }
}
```

**Reply by rhubarb-geek-nz** [reply]
Right, so they are just tags and don't actually indicate operating system in any formal or documented sense.

**Reply by ThomasNieto** [reply]
I tried to get an operation system manifest property added when additional properties were being proposed but all of them were rejected due to the breaking change.

**Reply by o-l-a-v** [reply]
Alright, but where is this behavior decided, documented, standardized? Why don't other PowerShell tools handling packages care about these tags / manifest properties?

PSResourceGet does not natively or optionally care about OS support though. Neither does `Microsoft.PowerShell.Core\Import-Module`, it blindly tries to import modules.

For a better user experience and friendlyness, I think PSResourceGet should check tags from the API vs. the host running the commands, and `Import-Module` should check the manifest / `.psd1`.

An example of current behavior: One can install and import `Microsoft.WinGet.Client` on Linux, but not use it.

```pwsh
PS /home/pop> Install-PSResource -Repository 'PSGallery' -Name 'Microsoft.WinGet.Client' -Scope CurrentUser -TrustRepository
PS /home/pop> Import-Module -Name 'Microsoft.WinGet.Client'                                                                 
PS /home/pop> Find-WinGetPackage -ID 'Microsoft.PowerShell'
&: The term 'WinGet' is not recognized as a name of a cmdlet, function, script file, or executable program. Check the spelling of the name, or if a path was
included, verify that the path is correct and try again.
PS /home/pop> $PSVersionTable                                                                                           

Name                           Value
----                           -----
PSVersion                      7.4.6
PSEdition                      Core
GitCommitId                    7.4.6
OS                             Pop!_OS 22.04 LTS
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

PS /home/pop> 
```

It should've been stopped at `Install-PSResource` already. And on `Import-Module`.

**Reply by rhubarb-geek-nz** [reply]
> Alright, but where is this behavior decided, documented, standardized? Why don't other PowerShell tools handling packages care about these tags / manifest properties?

[Microsoft.WinGet.Client.psd1](https://www.powershellgallery.com/packages/Microsoft.WinGet.Client/0.2.1/Content/Microsoft.WinGet.Client.psd1) does have both 

```
CompatiblePSEditions = 'Core'
```

and

```
        # Tags applied to this module. These help with module discovery in online galleries.
        Tags = @(
            'CrescendoBuilt',
            'PSEdition_Desktop',
            'PSEdition_Core',
            'Windows',
            'WindowsPackageManager',
            'WinGet'
        )
```

As the comment says the tags are for module discovery, not actual runtime behaviour.

**Reply by o-l-a-v** [reply]
Thanks for pointing that out. Then I guess this feature request is more about making it a runtime behavior for consistency and user friendliness.

**Reply by rhubarb-geek-nz** [reply]
> Then I guess this feature request is more about making it a runtime behavior

The RID format would be far more useful than those tags. Are we really in a situation where additional root manifest entries cannot be added?

**Reply by o-l-a-v** [reply]
If new root manifest entries can be added it could hopefully be done without breaking backwards compatibility. Lets keep current behavior for old (and new) manifests not utilizing new entries, but validate host OS vs. info in manifest for manifests that do?

**Comment by ThomasNieto** [comment]
Here is the issue I logged a while back for adding operating system to module manifest #9784 and related #3751.

**Reply by rhubarb-geek-nz** [reply]
> Here is the issue I logged a while back for adding operating system to module manifest #9784 and related #3751.

Looks like they were abandoned due to inactivity rather than rejection. The reason I would like RID rather than just operating system name is because I have a [module that is very specific](https://www.powershellgallery.com/packages/rhubarb-geek-nz.SQLiteConnection/1.0.119) about what it supports due to its native assemblies.

**Reply by ThomasNieto** [reply]
Its been a long time since these issues where logged and most of the discussions occurred outside of the Jira I logged. I'm sure there is a community call reference to it somewhere but here are the other RFC and issues relating to OS.

https://github.com/PowerShell/PowerShell/issues/10251
https://github.com/PowerShell/PowerShell-RFC/pull/122#issuecomment-460419644

**Reply by fMichaleczek** [reply]
@rhubarb-geek-nz Any change to the psd1 format would break backwards compatibility with PowerShell 5.1. So nobody wants to touch it. If we want something more, we need a new format and nobody has managed to find a consensus on the need.

I would need a new module format for the [Browser](https://github.com/dotnet/designs/blob/main/accepted/2023/net8.0-browser-tfm.md) port.  JavaScript Object Notation (JSON) should be better when hosted in a javascript engine ! Dll files do not go well through security network devices with content inspections (like proxy) , the dotnet team had to create a dedicated CIL format to solve the situation https://github.com/dotnet/runtime/blob/main/docs/design/mono/webcil.md

This is the right use case for developing a new specification.

![image](https://github.com/user-attachments/assets/927a9e2e-bcc9-4e30-a178-ffd5e6b6f733)

![image](https://github.com/user-attachments/assets/c59e0f82-de46-4662-9af9-fd761156e373)

![image](https://github.com/user-attachments/assets/9cf6796b-265a-4130-8a24-1a532be36847)
